python/tk_multi_workfiles/entity_models/deferred_model.py

```diff
@@ -0,0 +1,588 @@
+        """
+        # Basic sanity check that we do have a deferred query, so we don't have
+        # to test it everywhere.
+        if not deferred_query:
+            raise ValueError("A non empty deferred query is required")
+
+        self._deferred_query = deferred_query
+        self._task_step_icons = {}
+        super(ShotgunDeferredEntityModel, self).__init__(
+            entity_type,
+            filters,
+            hierarchy,
+            fields,
+            *args,
+            **kwargs
+        )
+        self._deferred_cache = ShotgunDataHandlerCache()
+        self._deferred_models = {}
+
+    @property
+    def deferred_query(self):
+        """
+        :returns: The deferred query for this model, if any.
+        """
+        return self._deferred_query
+
+    @property
+    def represents_tasks(self):
+        """
+        :returns: True if this model represents Tasks through deferred query.
+        """
+        return self._deferred_query["entity_type"] == "Task"
+
+    @property
+    def supports_step_filtering(self):
+        """
+        :returns: True if Step filtering can be used with this model
+        """
+        # If don't have steps in the fields we query from Shotgun, we assume
+        # step filtering should be disabled.
+        return "step" in self._deferred_query["hierarchy"]
+
+    def async_refresh(self):
+        """
+        Trigger an asynchronous refresh of the model
+        """
+        super(ShotgunDeferredEntityModel, self).async_refresh()
+        # Refresh our deferred cache
+        # Get the full list of uids
+        uids = [uid for uid in self._deferred_cache.uids]
+        # Retrieve parents for these uids
+        affected_parents = set()
+        for uid in uids:
+            item = self._get_item_by_unique_id(uid)
+            if item and item.parent():
+                # Let the parent re-fetch children
+                item.parent().setData(False, self._SG_ITEM_FETCHED_MORE)
+
+    def load_and_refresh(self, extra_filters=None):
+        """
+        Load the data for this model and post a refresh.
+
+        :param extra_filters: A list of additional Shotgun filters which are added
+                              to the initial filters.
+        """
+        self._extra_filters = extra_filters
+        # Extra filter is not applied to the model containing top nodes
+        # for deferred queries.
+        self._load_data(
+            self._entity_type,
+            self._original_filters,
+            self._hierarchy,
+            self._fields
+        )
+        self.async_refresh()
+
+    def ensure_data_for_context(self, context):
+        """
+        Ensure the data is loaded for the given context.
+
+        This is typically used to load data for the current Toolkit context and
+        select a matching item in the tree.
+
+        :param context: A Toolkit context.
+        """
+        if not context:
```

Comment:
We need to talk about moving this (or not) in sgutils. For the `ShotgunExtendedEntityModel`, there is some extra logic to use `context.task` or `context.entity`, it seems better to let the model deal with this logic, rather than doing it outside of it. And using a `context` here helps to understand the scope of this method.